src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/failover/AuroraTopologyService.java

```diff
@@ -257,33 +253,30 @@ private boolean gatherPerfMetrics(PropertySet props) {
    */
   private ClusterTopologyInfo processQueryResults(ResultSet resultSet)
       throws SQLException {
-    int writerCount = 0;
 
     List<HostInfo> hosts = new ArrayList<>();
+    List<HostInfo> writers = new ArrayList<>();
     while (resultSet.next()) {
+      HostInfo currentHost = createHost(resultSet);
+
       if (!WRITER_SESSION_ID.equalsIgnoreCase(resultSet.getString(FIELD_SESSION_ID))) {
-        hosts.add(createHost(resultSet));
+        hosts.add(currentHost);
         continue;
       }
 
-      if (writerCount == 0) {
-        // store the first writer to its expected position [0]
-        hosts.add(
-            FailoverConnectionPlugin.WRITER_CONNECTION_INDEX,
-            createHost(resultSet));
-      } else {
-        // append other writers, if any, to the end of the host list
-        hosts.add(createHost(resultSet));
-      }
-      writerCount++;
+      writers.add(currentHost);
     }
 
-    if (writerCount == 0) {
+    if (writers.size() == 0) {
       this.log.logError(Messages.getString("AuroraTopologyService.3"));
       hosts.clear();
+    } else {
+      // Store the first writer to its expected position [0]. If there are other writers or stale records, ignore them.
+      // ResultSet and writers list is already ordered by latest updated after querying with RETRIEVE_TOPOLOGY_SQL.
+      hosts.add(FailoverConnectionPlugin.WRITER_CONNECTION_INDEX, writers.get(0));
```

Comment:
writers.get(0) may contain not the latest writer node. I think we still need to sort writers collection by last updated date.